Refinement Ratio of 4 Support in Nodal Solver

WeiqunZhang opened the pull request:
This adds refinement ratio of 4 support in non-EB nodal linear solver.
Roundoff errors are expected for multi-level solves with a refinement ratio
of 2.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

WeiqunZhang committed:
Refinement Ratio of 4 Support in Nodal Solver

This adds refinement ratio of 4 support in non-EB nodal linear solver.
Roundoff errors are expected for multi-level solves with a refinement ratio
of 2.

WeiqunZhang commented:
@cgilet Will you be able to test this with IAMR?

cgilet commented:
I'm getting an out of bounds error in setBndryValues during the MAC projection. Is that supposed to be working with ref_ratio==4?

Here's part of the backtrace:
```
amrex::InterpBndryData::BndryValuesDoIt(amrex::BndryRegister&, int, amrex::MultiFab const*, int, int, int, amrex::IntVect const&, amrex::BCRec const*, int) at /home/candace/CCSE/IAMR_cleanup/Exec/run3d/../../../amrex_weiqun/amrex/Src/Base/AMReX_Array4.H:96
 (inlined by) ?? at /home/candace/CCSE/IAMR_cleanup/Exec/run3d/../../../amrex_weiqun/amrex/Src/Boundary/AMReX_InterpBndryData_3D_K.H:34
 (inlined by) amrex::InterpBndryData::BndryValuesDoIt(amrex::BndryRegister&, int, amrex::MultiFab const*, int, int, int, amrex::IntVect const&, amrex::BCRec const*, int) at /home/candace/CCSE/IAMR_cleanup/Exec/run3d/../../../amrex_weiqun/amrex/Src/Boundary/AMReX_InterpBndryData.cpp:225

amrex::InterpBndryData::setBndryValues(amrex::BndryRegister&, int, amrex::MultiFab const&, int, int, int, amrex::IntVect const&, amrex::BCRec const&, int) at /home/candace/CCSE/IAMR_cleanup/Exec/run3d/../../../amrex_weiqun/amrex/Src/Boundary/AMReX_InterpBndryData.cpp:144
```

WeiqunZhang commented:
MAC projection is supposed to work.  Could you push your IAMR change somewhere so that I can reproduce the error you see.

WeiqunZhang commented:
The problem is in `MacProjector::setDomainBC`.  It should not call `setLevelBC`, which needs to know the amr refinement ratio.  We did it this way to make `MacProjector` more user friendly.  But unfortunately it breaks for refinement ratio of 4.  So we will have to change this, and it will be the user's responsibility to call `setLevelBC`.

WeiqunZhang commented:
Actually, there is a way to keep the current behavior of calling `setLeveBC` implicitly.

WeiqunZhang committed:
MacProjector refinement ratio of 4 bug

Move setLevelBC to project and call it only if it has not been called.  This
fixes a refinement ratio of 4 bug, because setLevelBC should not be called
before coarse/fine ratio is set.
